velox/functions/lib/TDigest.h

```diff
@@ -0,0 +1,484 @@
+        continue;
+      }
+      auto wi = weights_[i];
+      auto mi = means_[i];
+      auto j = i;
+      for (;;) {
+        auto k = positions[j];
+        if (k == i) {
+          break;
+        }
+        weights_[j] = weights_[k];
+        means_[j] = means_[k];
+        positions[j] = j;
+        j = k;
+      }
+      weights_[j] = wi;
+      means_[j] = mi;
+      positions[j] = j;
+    }
+    VELOX_DCHECK(std::is_sorted(means_.begin(), means_.end()));
+  }
+  merge(compression, weights_.data(), means_.data(), weights_.size());
+}
+
+template <typename A>
+void TDigest<A>::merge(
+    double compression,
+    const double* weights,
+    const double* means,
+    int count) {
+  VELOX_CHECK_GT(count, 0);
+  VELOX_CHECK_GE(weights_.size(), count);
+  if (reverseCompress_) {
+    // Run the merge in reverse every other merge to avoid left-to-right
+    // bias.
+    mergeImpl<true>(compression, weights, means, count);
+  } else {
+    mergeImpl<false>(compression, weights, means, count);
+  }
+  reverseCompress_ = !reverseCompress_;
+}
+
+template <typename A>
+template <bool kReverse>
+void TDigest<A>::mergeImpl(
+    double compression,
+    const double* weights,
+    const double* means,
+    int count) {
+  const auto totalWeight = std::accumulate(weights, weights + count, 0.0);
+  const auto invTotalWeight = 1 / totalWeight;
+  const auto normalizer =
+      (4 * std::log(totalWeight / compression) + 24) / compression;
+  auto maxSize = [normalizer](double q) { return q * (1 - q) * normalizer; };
+  double weightSoFar = 0;
+  numMerged_ = 0;
+  const int begin = kReverse ? count - 1 : 0;
+  auto notEnd = [&](auto i) INLINE_LAMBDA {
+    if constexpr (kReverse) {
+      return i >= 0;
+    } else {
+      return i < count;
+    }
+  };
+  constexpr int kStep = kReverse ? -1 : 1;
+  int j = begin;
+  weights_[j] = weights[begin];
+  means_[j] = means[begin];
+  for (int i = begin + kStep; notEnd(i); i += kStep) {
+    auto proposedWeight = weights_[j] + weights[i];
+    auto q0 = weightSoFar * invTotalWeight;
+    auto q2 = (weightSoFar + proposedWeight) * invTotalWeight;
+    if (proposedWeight <= totalWeight * std::min(maxSize(q0), maxSize(q2))) {
+      weights_[j] += weights[i];
+      means_[j] += (means[i] - means_[j]) * weights[i] / weights_[j];
+    } else {
+      weightSoFar += weights_[j];
+      ++numMerged_;
+      j += kStep;
+      weights_[j] = weights[i];
+      means_[j] = means[i];
+    }
+  }
+  weightSoFar += weights_[j];
+  ++numMerged_;
+  VELOX_CHECK_LT(std::abs(weightSoFar - totalWeight), kEpsilon);
+  if constexpr (kReverse) {
+    std::copy(weights_.begin() + j, weights_.end(), weights_.begin());
+    std::copy(means_.begin() + j, means_.end(), means_.begin());
+  }
+  weights_.resize(numMerged_);
+  means_.resize(numMerged_);
+  min_ = std::min(min_, means_.front());
+  max_ = std::max(max_, means_.back());
+}
+
+template <typename A>
+double TDigest<A>::estimateQuantile(double quantile) const {
```

Comment:
Hi @Yuhta, could you add the comments at corresponding branches in the Java's getQuantile() method? I found those comments helpful for understanding the math here. Thanks!